### DISCUSSION TITLE
504 error | serverless function | PRO member | maxDuration 5 minute

### DISCUSSION THREAD

**Discussion by space1worm** [vercel/community#4248] [Help] [answered]
### Summary

**Nextjs version 13.4.19
Using App Directory
Pro member**

I am encountering a 504 error on my request. I'm attempting to await some functionality from a serverless function that takes roughly 3 minutes to complete. I have upgraded my plan to PRO. Additionally, I've exported the maxDuration variable from **src/app/api/generators/article/refine/route.ts** as follows

**route.ts**
```
export const maxDuration = 300;

export async function POST(request: NextRequest) {
  const payload = (await request.json()) as GetArticle;

  return GeneratorsAPI.getInstance().getRefinedArticle(payload);
}
```

I've also provided a vercel.json file, which essentially instructs Next.js to increase the serverless execution time to 300 seconds. Furthermore, I've increased the Axios timeout parameter. However, I'm still receiving a 504 error after 1 minute of loading time.

**vercel.json in root directory e.i /vercel.json**
```
{
  "functions": {
    "app/api/generators/**/*.ts": {
      "memory": 3008,
      "maxDuration": 300
    }
  }
}

```

I have also increased axios timeout on client and be as well

**GeneratorsAPI.ts**

```
 const response = await axios.post(this.REFINED, payload, {
        headers: this.headers,
        timeout: 300000, // 5 minutes
      });
```

**actions.ts**

```
const response = await axios.post(API_CLIENT_ENDPOINTS.refinedArticle, data, {
        responseType: "json",
        timeout: 300000, // 5 minutes
      });
```

I can also confirm that on vercel I indeed can see that **maxDuration** is increased for my **serverless functions** to 5 minute

Please assist me with this issue.

### Example

https://smartapp-three.vercel.app/tools/article-generator

### Steps to Reproduce

Go to -> https://smartapp-three.vercel.app/tools/article-generator

Type anything in Email field -> wait around 1 minute.

if ur lucky u will see outlines and you can click one of them and then again resubmit form and you will see 504 error.
If not you will see immediately 504 error.  in network logs

**Comment by Felixljr** [comment]
Double check your vercel.json set up. It's structured well but if it's in root then your path should start from src/ I had a similar issue bc I followed the docs but I guess their example was for a setup without src/. Also this was the only file/configuration I needed to alter maxDuration. Hope this helps.

**Reply by space1worm** [reply]
unfortunately it didn't help... I do see that my functions maxDuration is 5 minutes on VERCEL functions page but when I create request to that API after 1 minute it throws 504 Gateway Timeout error ...

**Reply by Felixljr** [reply]
The only other suggestion I can make if you haven't tried yet is to remove the wildcards from your path and hard code it directly to the route file. If this works then you might just have to adjust where you are placing "/**/*.ts" in your path; or leave it hard coded. If it doesn't work I'm sorry that's all I got, this is as far as I had to go to resolve my similar issue.

**Reply by space1worm** [reply]
The thing is I created new project, just to test everything from scratch with  ```npx create-next-app@latest``` without src directory

I created endpoint api/hello/route.ts and accordingly vercel.json

```
{
  "functions": {
    "app/api/hello/route.ts": {
      "memory": 3008,
      "maxDuration": 300
    }
  }
}

```

I see in vercel that hello route has ```5 minute maxDuration``` but when I send request again after 1 minute dies and sends back Gateway Timeout error but function itself resolves and returns response after some time in vercel logs...

**Reply by hnykda** [reply]
I have exactly the same issue: https://github.com/vercel/next.js/issues/56447#issuecomment-1749530759

**Reply by mmassaly** [reply]
Hello I have the same issue with my page. It sows / 300s but I get 504 TImeout after 15s. They have scammed us guys.

**Reply by leerob** [reply]
Can you share the code for your function? Are you on a paid Pro plan and not a Pro trial?

**Comment by amyegan** [comment]
I tried to replicate the problem, but the only way I'm able to get a timeout error is if my function really does exceed the 300s limit.
<img width="1291" alt="image" src="https://github.com/vercel/community/assets/8247105/68979878-3e78-428c-911f-3d88236007b8">

I tried it with `maxDuration` in my function code and I tried it with the `functions` property in my `vercel.json` file. Both had the same result. Do you have a small [MRE](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that I can use to get a better look at the problem?

**Reply by hnykda** [reply]
Hey. Thanks for looking into this. One guy posted an example here: https://github.com/yakhinvadim/vercel-maxduration (from https://github.com/vercel/next.js/issues/56447#issue-1926982818)

**Reply by space1worm** [reply]
Hey, yes sure!

you can check it more detailed description here -> https://github.com/vercel/next.js/issues/56413
You can also check codebase on here -> https://github.com/space1worm/serverless-timeout-test

**Reply by amyegan** [reply]
Thanks for sharing these examples. I still wasn't able to reproduce the early timeout problem on my Pro account though 😕 

<img width="358" alt="image" src="https://github.com/vercel/community/assets/8247105/0c63d1fa-b15c-46ab-8c2d-c506fe157d86">

<img width="355" alt="image" src="https://github.com/vercel/community/assets/8247105/7e5c43c5-9d04-4920-8e3b-647934034313">

**Reply by space1worm** [reply]
We could arrange call, so I could show whats going on exactly on my pro account. If that will help you to debug problem better. 😕

**Comment by dlibis** [comment]
Im also having issues, with this, would love some help. 
before the change, 60 seconds was sufficient, but not i cant override the default 15 seconds.

**Reply by space1worm** [reply]
You have to be a PRO member to do it. 

For free tier you can have 15s execution time Max.

**Comment by 8ctopotamus** [comment]
Finally figured out the 504 Timeout issue in my project! 

The issue was is that we have `/app/some-route/[slug]/page.tsx` templates that are trying to statically generate hundreds of pages, and [Vercel's new time limit](https://vercel.com/changelog/serverless-functions-can-now-run-up-to-5-minutes) now caps us at 15s. It was the step of fetching all of the paths in `generateStaticParams` that was taking longer than 15s. This had worked fine up until the start of October 2023.

The fix is to stop using `generateStaticParams` in these templates and instead dynamically generate the page upon request. [To do this in the App Router API](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams), you simply add `export const dynamicParams = true` at the top of the page.tsx .

```tsx

// allows next to dynamically generate the page upon request
export const dynamicParams = true

// This now goes over the 15s time limit if your path-fetching call takes a long time
// export async function generateStaticParams() {
//   return await getAllPaths()
// }

// rest of page.tsx...

```

**Reply by space1worm** [reply]
Unfortunately thats not the case for my project.

We have a POST api handler which has timeout issues.

**Reply by 8ctopotamus** [reply]
Gotcha, @space1worm. Different situation. I see in your original post that you tried `maxDuration`. This setting also seemed to do nothing for me, even though we are on a Pro account.

**Reply by space1worm** [reply]
Yeah... thats very strange.

Keep in mind that ```export const maxDuration``` works technically only after Next version 13.5.* as I remember.

In our case we are using next@latest but still nothing..

**Reply by hnykda** [reply]
I am in the same boat as space1worm, no staticParams and what not. Just ol' plain API route. I am on 13.5 as well.

**Reply by fmsf** [reply]
Had the same problem with slugs, but don't have `staticParams` , adding `export const maxDuration = 300;` to my `page.tsx` file that receives the traffic from the slug route worked for me. i.e.: here: `src/app/api/foo/[[...query]]/page.tsx`

**Comment by swarnava** [comment, marked as answer]
Hi,

We have pushed a fix for this issue. You should be able to use function longer than 60 seconds now. Let us know if you have any question.

**Reply by WesselKoorn** [reply]
Hi @swarnava, is the fix live already? I'm still running in exactly the same issue described before.

I export the `maxDuration` in my api endpoint (`pages/api/my-resource/[id].ts`):
```
import { NextApiRequest, NextApiResponse } from "next";

export const maxDuration = 300; // 5 minutes.

export default async function longRunningTask(
  req: NextApiRequest,
  res: NextApiResponse
) {
...
```

I'm running Next 13.5.4:

![Screenshot 2023-10-09 at 11 57 47](https://github.com/vercel/community/assets/9317420/f859a2da-dbce-42f0-9e40-72047947ba88)

But the set duration limit is not recognized:

![Screenshot 2023-10-09 at 11 59 28](https://github.com/vercel/community/assets/9317420/c395c8e6-411e-4e7b-9cc0-7de8cbefd475)

Am I missing something?

**Reply by space1worm** [reply]
You have to be on a PRO plan.

15s is maximum for Hobby plan (Free)

**Reply by WesselKoorn** [reply]
Sorry I didn’t mention, but yes I’m on a PRO plan.

**Reply by space1worm** [reply]
you have to export either GET or POST or PUT or DELETE and it will be pickedup by Vercel automatically. 

You can't just use 'randomNameForMyFunction' for API handler as I know.

**Reply by WesselKoorn** [reply]
UPDATE: My bad, I didn't read [the docs](https://hookify-m33dyoj6x-through-the-roof-innovations.vercel.app/login) correctly 🤦‍♂️. I was looking at `Next.js (/app)` while I'm working with `Next.js (/pages)`, so I need to export a config object instead:

```
import { NextApiRequest, NextApiResponse } from "next";

export const config = {
  maxDuration: 300, // 5 minutes.
};

export default async function longRunningTask(
  req: NextApiRequest,
  res: NextApiResponse
) {
...
```

I wasn't aware of the new App Router yet, sorry for the confusion.

**Reply by faye1225** [reply]
<img width="963" alt="image" src="https://github.com/vercel/community/assets/56664035/c9f3ac08-cd83-42b3-b2ab-54f5c7518d6c">
I have an API request that takes a long time to process. I set maxDuration, but it doesn't take effect. Below are my relevant configurations and code.

```
export const runtime = 'edge'
export const maxDuration = 300
export async function POST(req: Request) {
    const uploadJson = await fetch(...)
...
    return NextResponse.json(uploadJson)
}
```
"next": "13.5.5-canary.17" and I’m on a PRO plan

**Reply by space1worm** [reply]
@faye1225  Your runtime is 'edge' thats why you are experiencing that issue, remove that. 

On the other hand if you really want to use 'edge' and u have special usecase for it then consider using this approach -> https://vercel.com/docs/functions/edge-functions/streaming

<img width="820" alt="Screenshot 2023-10-16 at 16 31 46" src="https://github.com/vercel/community/assets/84189246/872aaae3-4db2-419e-bdd8-c780dcd0b4cb">

**Reply by dlibis** [reply]
@space1worm leads to 404 page

**Reply by amyegan** [reply]
Like space1worm said, this is an Edge Function which has different configuration and duration limits from Serverless Functions. An Edge Function has a [maximum initial response time](https://vercel.com/docs/functions/edge-functions/limitations#maximum-initial-response-time) of 25 seconds, but it can continue streaming a response after the initial response. You can find info about streaming data at https://vercel.com/docs/functions/streaming

If you'd rather make it a Serverless Function with a [`maxDuration`](https://vercel.com/docs/functions/serverless-functions/runtimes#maxduration) value, just remove `export const runtime = 'edge'` from your code. One thing to note is the optional higher duration limit is only available on Pro plans. The Hobby limit for Serverless Function duration is 10 seconds.

**Reply by faye1225** [reply]
@amyegan @space1worm 
Thank you for your replies, I have solved the problem without any problems. I have summarized what is needed
- next version needs to be above  13.5.4 
- needs to be pro plan
- remove `export const runtime = 'edge'`. Or change to `export const runtime = 'nodejs'`


<img width="576" alt="image" src="https://github.com/vercel/community/assets/56664035/b9d0e387-414c-4c04-b6a4-105203f9d0c6">
<img width="640" alt="image" src="https://github.com/vercel/community/assets/56664035/a99f2ee3-c219-4361-8da5-449246b90ebe">

**Comment by dlibis** [comment]
how do you handle this globally for ISR pages? 
my current solution is to put maxDuration const on every page.

**Reply by space1worm** [reply]
as docs states, basically for generating pages (static or isr) Next.js uses serverless functions for it. It means that if this process (fetching data) takes more time than DEFAULT (15 for hobby, 1m for pro) then you need to manually increase it otherwise it will fail.

**Reply by swarnava** [reply]
Here is an example: https://vercel.com/docs/projects/project-configuration#functions-property-with-isr

**Reply by dlibis** [reply]
I tired to do it, for example, 
my page is located in : "app/(main)/(overview)/page.tsx", so i put in the config 
`{
  "functions": {
    "app/**/**/*": {
      "maxDuration": 10
    }
  }
}`

and theres no effect, the duration is the default. (im on pro plan)

**Comment by dinakartumu** [comment]
Does any one know how to increase maxDuration for project using redwood js framework. i tired `{
  "functions": {
    "api/graphql": {
      "maxDuration": 120
    }
  }
}
` adding vercel.json and my builds are failing with 'Error: The pattern "api/graphql" defined in `functions` doesn't match any Serverless Functions inside the `api` directory.'

**Reply by amyegan** [reply]
Hi @dinakartumu! I just responded to your other comment on https://github.com/vercel/community/discussions/4391#discussioncomment-7256253

**Comment by prageeth** [comment]
I'm still have the same issue. I'm on a Pro plan, with `export const maxDuration = 300;` set but I'm still getting the 15s timeout.
I've not set any other configurations for these endpoints.

<img width="729" alt="image" src="https://github.com/vercel/community/assets/230793/b5556813-83d3-4615-982d-c70b73869a9b">

**Reply by swarnava** [reply]
It looks like you have a ticket with us. We will take it from there. Thank you!

**Reply by ferealqq** [reply]
Same, the weird thing is that this same endpoint worked whilst I had the free trial on. But when I had to pay the for the proplan it stopped working?


![image](https://github.com/vercel/community/assets/22598325/f1669214-327c-4e7f-ae50-9ffc28ca7778)

**Reply by amyegan** [reply]
Hi @ferealqq. How is `maxDuration` configured for this function? If you can share code snippets or [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), that will give me a better idea of what may be causing the problem.

**Reply by ferealqq** [reply]
```javascript
// [trpc].ts....
export const maxDuration = 180;

export const runtime = "nodejs";
```

Next version: 13.5.5

Haven't had time to make a reproducible example i'll try to do it later this week prob tomorrow. Have there been more of these cases recently? What should I check for?

**Reply by prageeth** [reply]
@ferealqq What worked for me in the end was to use as I was using the page router.
```
export const config = {
  maxDuration: 300,
};
```

**Reply by swarnava** [reply]
Note that the above instruction may differ based on App Router or Page Router you are using with Next.js: https://vercel.com/docs/functions/configuring-functions/duration#node.js,-next.js-13.5-or-higher,-sveltekit,-and-remix

**Reply by darksipu** [reply]
I'm having the same issue. I'm on Next 14.04, and I've got this code in pages/api/functionName.js

`export const config = {
  maxDuration: 300,
}`

The logs on Vercel still show me a max run of 15s. I'm on the Pro account trial.

I also tried configuring it globally and had no luck.

**Reply by makluganteng** [reply]
Hi i also have the same error where i added `export const maxDuration = 300` im also using pro plan but i cant seem to make it 300s it still shows 15 second :((.

**Reply by swarnava** [reply]
Please create a separate thread for this, Thank you!
